incubator/hnc/pkg/controllers/object_controller.go

```diff
@@ -104,68 +103,93 @@ func (r *ObjectReconciler) SyncNamespace(ctx context.Context, log logr.Logger, n
 // update deletes this object if it's an obsolete copy, and otherwise ensures it's been propagated
 // to any child namespaces.
 func (r *ObjectReconciler) update(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) error {
-	// Make sure this object is correct and supposed to propagate. If not, delete it or mark it as
-	// modified. This should trigger one more reconciliation cycle and we'll get the children then.
-	if correct, err := r.ensureCorrectAncestry(ctx, log, inst); !correct || err != nil {
-		log.Info("Incorrect will not propagate")
-		return err
+	// Make sure this object is correct and supposed to propagate. If not, delete it.
+	srcInst := r.getSourceInst(ctx, log, inst)
+	dests, good := r.syncWithForest(ctx, log, inst, srcInst)
+	if !good {
+		return r.Delete(ctx, inst)
 	}
 
 	// Make sure this gets propagated to any children that exist
-	return r.propagate(ctx, log, inst)
+	return r.propagate(ctx, log, inst, dests)
 }
 
-// ensureCorrectAncestry checks to see if the given object is correct. It returns true if
-// it's correct and ready to be propagated. The incorrect object will be deleted
-// if it's obsolete; or marked as modified if it's changed.
-func (r *ObjectReconciler) ensureCorrectAncestry(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, error) {
-	// Find what we expect the source to be.
-	missing, src, err := r.getSourceInst(ctx, log, inst)
-	if err != nil {
-		return false, err
+// syncWithForest syncs the object instance with the in-memory forest. This is the only place we
+// should lock the forest, so this fn needs to return everything relevant for the rest of the
+// reconciler. It returns true if the object instance is good and shouldn't be deleted. It will
```

Comment:
We should also add that this method doesn't (or more accurately, shouldn't) contact the apiserver since that's a slow operation and everything will block on the lock being held.